### PR TITLE
[TEAM6] Final PR

### PR DESCRIPTION
**Pull request by MattScottEEE**
# Summary

Added treaty strategies, messaging + behavioural changes

## Test Plan

Rigorous testing in run - agents are fair and honest

### CODE REVIEW

File: config.json
Comment (JaafarRammal) [inline config.json:3]: Prefer to leave config as is!

File: pkg/agents/team6/behaviours.go
Comment (JaafarRammal) [inline pkg/agents/team6/behaviours.go:24]: Just checking that `float64(a.maxFloorGuess)` is never 0?

File: pkg/agents/team6/treaties.go
Comment (JaafarRammal) [inline pkg/agents/team6/treaties.go:30]: I see you set `params.r` to never be 0 in the other function, but just confirming that `r` will never be set to 0 or 1 during sim?

**Review by JaafarRammal — CHANGES_REQUESTED** [review]
Code LGTM, a few comments left.
Would be good to revert config file.
I think `func (a *CustomAgent6) considerTreaty(t *messages.Treaty) bool` in `pkg/agents/team6/treaties.go` can be visually improved as it's very nested

File: pkg/agents/team6/communication.go
Comment (jzzheng22) [inline pkg/agents/team6/communication.go:109]:
```suggestion
		reply = currentHP >= levels.weakLevel
```

File: pkg/agents/team6/communication.go
Comment (jzzheng22) [inline pkg/agents/team6/communication.go:116]:
```suggestion
		reply = currentHP >= levels.strongLevel
```

File: pkg/agents/team6/behaviours.go
Comment (MattScottEEE) [reply pkg/agents/team6/behaviours.go:24]: Nice spot - we always initialise as our current floor and then update if we get a score bigger - not possible to get 0

File: config.json
Comment (MattScottEEE) [reply config.json:3]: Yes that's on me - I thought I updated from main but perhaps not!

File: pkg/agents/team6/communication.go
Comment (jzzheng22) [inline pkg/agents/team6/communication.go:154]: ??

Dead code
Just pass `false` directly into `Reply()`
And you don't need the if statement, since the if block will never execute, only the else

File: pkg/agents/team6/treaties.go
Comment (MattScottEEE) [reply pkg/agents/team6/treaties.go:30]: Again, fair spot - these params are static (and we have a special case for when we have `params.r == 1`)

File: pkg/agents/team6/config.go
Comment (jzzheng22) [inline pkg/agents/team6/config.go:125]:
```suggestion
		longTermMemory:      make(memory, 0),
		shortTermMemory:     make(memory, 0),
```

File: pkg/agents/team6/behaviours.go
Comment (JaafarRammal) [reply pkg/agents/team6/behaviours.go:24]: Cool

**Comment by MattScottEEE** [comment]
> Code LGTM, a few comments left.
> Would be good to revert config file.
> I think `func (a *CustomAgent6) considerTreaty(t *messages.Treaty) bool` in `pkg/agents/team6/treaties.go` can be visually improved as it's very nested

Thanks for this - reverting the config + taking a look at this function:

overall, team think that, despite the nested-ness, it preserves readability instead of adding lots of sub-functions. Could I ask for a recommendation of what we might be able to do?

File: pkg/agents/team6/consumption.go
Comment (jzzheng22) [inline pkg/agents/team6/consumption.go:116]:
```suggestion
	}
	return maxAllowedFood
```

File: pkg/agents/team6/treaties.go
Comment (JaafarRammal) [reply pkg/agents/team6/treaties.go:30]: What I'm seeing is that if `!(socialMotive == "Altruist" || socialMotive == "Nacissist)"`, regardless of the value of `r`, this is reached. If `r` is set to 1 what is the behavior? (because you get `params.r/(1-params.r) = 1/0`)

In all fairness, the only time the calculation would be valid is when `(params.c*params.r)/params.g` is 1, you then get `x^(x/(1-x))`

If `r` never reaches an illegal value, feel free to resolve the convo!

File: pkg/agents/team6/memory.go
Comment (jzzheng22) [inline pkg/agents/team6/memory.go:29]:
```suggestion
	return a.prevFloor != a.Floor() && a.prevFloor != -1
```

File: pkg/agents/team6/treaties.go
Comment (jzzheng22) [inline pkg/agents/team6/treaties.go:94]: Don't compare the numbers - use the enum here instead.

File: pkg/agents/team6/treaties.go
Comment (jzzheng22) [inline pkg/agents/team6/treaties.go:110]: I suggest putting this into a function

File: pkg/agents/team6/treaties.go
Comment (jzzheng22) [inline pkg/agents/team6/treaties.go:123]: I suggest putting this into a function

File: pkg/agents/team6/treaties.go
Comment (jzzheng22) [inline pkg/agents/team6/treaties.go:150]: I suggest putting this into a function

File: pkg/agents/team6/treaties.go
Comment (jzzheng22) [inline pkg/agents/team6/treaties.go:162]: I suggest putting this into a function

File: pkg/agents/team6/treaties.go
Comment (jzzheng22) [inline pkg/agents/team6/treaties.go:179]: Function is too long - use functions to shorten it

File: config.json
Comment (jzzheng22) [inline config.json]: Did yuo change this file...?

File: pkg/agents/team6/treaties.go
Comment (jzzheng22) [inline pkg/agents/team6/treaties.go:252]: Don't need else, you can just return false

File: pkg/agents/team6/treaties.go
Comment (jzzheng22) [inline pkg/agents/team6/treaties.go:278]: Unsafe to use numbers, use the enum

Comment (jzzheng22) [reply pkg/agents/team6/treaties.go:278]: I'd also put this into a function

Comment (jzzheng22) [reply pkg/agents/team6/treaties.go:278]: Both of the above comments apply to the other case statements too

**Review by jzzheng22 — CHANGES_REQUESTED** [review]
Some of the code is unsafe - needs to be changed.

File: pkg/agents/team6/treaties.go
Comment (MattScottEEE) [reply pkg/agents/team6/treaties.go:30]: this is because the value of r is specific to the behaviour (initialised in config!) - testing for behaviour also tests for r :)

File: config.json
Comment (JaafarRammal) [reply config.json]: Latest file on main is:

```json
{
    "UseFoodPerAgentRatio": false,
    "FoodOnPlatform": 100,
    "FoodPerAgentRatio": 0,
    "Team1Agents": 2,
    "Team1Agents2": 2,
    "Team2Agents": 2,
    "Team3Agents": 2,
    "Team4Agents": 2,
    "Team5Agents": 2,
    "Team6Agents": 2,
    "Team7Agents": 2,
    "RandomAgents": 2,
    "AgentHP": 100,
    "AgentsPerFloor": 1,
    "TicksPerFloor": 10,
    "SimDays": 8,
    "ReshuffleDays": 1,
    "maxHP": 100,
    "weakLevel": 10,
    "width": 48,
    "tau": 15,
    "MaxFoodIntake": 60,
    "hpReqCToW": 5,
    "hpCritical": 3,
    "maxDayCritical": 3,
    "HPLossBase": 5,
    "HPLossSlope": 0.2,
    "LogMain": true,
    "LogStory": true,
    "SimTimeoutSeconds": 60
}
```

File: pkg/agents/team6/treaties.go
Comment (MattScottEEE) [reply pkg/agents/team6/treaties.go:30]: (reopening so you can see response)

Comment (mdubied) [reply pkg/agents/team6/treaties.go:30]: `r=1` is only the case for `socialMotive="Altruist"`. In this case, we go in the first `if` statement where we do not divide by `(1-r)`

File: pkg/agents/team6/communication.go
Comment (SaraFFdez) [inline pkg/agents/team6/communication.go:215]: Should you sign the treaty here too to increase the treaty count?

**Comment by JaafarRammal** [comment]
> overall, team think that, despite the nested-ness, it preserves readability instead of adding lots of sub-functions. Could I ask for a recommendation of what we might be able to do?

@MattScottEEE That's correct, it's actually readable, and I think it's the logs that got in the way. This was written 13 times (exactly same line), could a call wrapped to make it cleaner actually: 

`a.Log("I used considerTreatyUsingUtility", infra.Fields{"my floor:": a.Floor(), "my social motive": a.currBehaviour.string()})`

File: pkg/agents/team6/treaties.go
Comment (AdrianKoch3010) [reply pkg/agents/team6/treaties.go:179]: It is like a lookup table, putting the individual switch cases into the function is not going to improve readability

File: pkg/agents/team6/communication.go
Comment (MattScottEEE) [reply pkg/agents/team6/communication.go:215]: Let me check this one - cheers for spotting <3

File: pkg/agents/team6/communication.go
Comment (MattScottEEE) [reply pkg/agents/team6/communication.go:154]: False was a placeholder - updating now (`sorry for this :((( `)

File: pkg/agents/team6/treaties.go
Comment (jzzheng22) [reply pkg/agents/team6/treaties.go:179]: No, this is extremely difficult to read. 
The way you handle `Collectivist`, and `Selfish` should be in functions.
The way you handle `HP` and `AvailableFood` should be in functions.

File: pkg/agents/team6/communication.go
Comment (MattScottEEE) [reply pkg/agents/team6/communication.go:205]: We interpreted it two ways:

1.) either we sign a treaty then sign it off
2) or we receive a treaty and sign on their behalf

We assume that the former is more logical, assuming that the treaty response we receive is already signed

Comment (MattScottEEE) [reply pkg/agents/team6/communication.go:205]: (in which case, we shouldn't have to sign)

**Comment by MattScottEEE** [comment]
> > overall, team think that, despite the nested-ness, it preserves readability instead of adding lots of sub-functions. Could I ask for a recommendation of what we might be able to do?
> 
> @MattScottEEE That's correct, it's actually readable, and I think it's the logs that got in the way. This was written 13 times (exactly same line), could a call wrapped to make it cleaner actually:
> 
> ```go
> a.Log("I used considerTreatyUsingUtility", infra.Fields{"my floor:": a.Floor(), "my social motive": a.currBehaviour.string()})
> ```
> 
> Maybe it should go inside the `considerTreatyUsingUtility` function since this is when you are calling this log

Great suggestion - thanks for this

File: pkg/agents/team6/communication.go
Comment (SaraFFdez) [reply pkg/agents/team6/communication.go:205]: Fair

File: pkg/agents/team6/treaties.go
Comment (jzzheng22) [reply pkg/agents/team6/treaties.go:179]: This is largely resolved by #289

File: pkg/agents/team6/treaties.go
Comment (jzzheng22) [reply pkg/agents/team6/treaties.go:278]: This is largely resolved by #289

File: pkg/agents/team6/communication.go
Comment (mdubied) [reply pkg/agents/team6/communication.go]: We updated it. It does not always return false now

File: pkg/agents/team6/treaties.go
Comment (jzzheng22) [inline pkg/agents/team6/treaties.go:154]:
```suggestion
```

File: pkg/agents/team6/treaties.go
Comment (jzzheng22) [inline pkg/agents/team6/treaties.go:190]: Would personally move the long term benefit stuff into the else statement since you don't need it unless you enter the else.

File: pkg/agents/team6/treaties.go
Comment (jzzheng22) [inline pkg/agents/team6/treaties.go:210]: optional, but maybe consider whether these comments are necessary here

File: pkg/agents/team6/treaties.go
Comment (jzzheng22) [inline pkg/agents/team6/treaties.go:271]: Up to you here, but I would use the ConditionType and RequestType, not the numbers

File: pkg/agents/team6/treaties.go
Comment (jzzheng22) [inline pkg/agents/team6/treaties.go:274]: Up to you here, but I would use the ConditionType and RequestType, not the numbers

File: pkg/agents/team6/treaties.go
Comment (jzzheng22) [inline pkg/agents/team6/treaties.go:281]: Up to you here, but I would use the ConditionType and RequestType, not the numbers

File: pkg/agents/team6/treaties.go
Comment (jzzheng22) [inline pkg/agents/team6/treaties.go:284]: Up to you here, but I would use the ConditionType and RequestType, not the numbers

**Review by jzzheng22 — APPROVED** [review]
Have a few suggestions but generally they're up to you.

Will be up for a little longer if you do decide to change those.

Would prefer if you wait for a second opinion before merging

File: pkg/agents/team6/behaviours.go
Comment (MattScottEEE) [reply pkg/agents/team6/behaviours.go:44]: Yes - string locks to 1 of 4 options in a range, we want to check the behaviour change, not just the numerical change

File: pkg/agents/team6/treaties.go
Comment (mdubied) [reply pkg/agents/team6/treaties.go:210]: We'll leave it

File: pkg/agents/team6/treaties.go
Comment (MattScottEEE) [reply pkg/agents/team6/treaties.go:271]: Agree fully - makes the code secure. Changing all

File: pkg/agents/team6/treaties.go
Comment (mdubied) [reply pkg/agents/team6/treaties.go:281]: done

File: pkg/agents/team6/treaties.go
Comment (MattScottEEE) [reply pkg/agents/team6/treaties.go]: nice spot - done :)

Comment (mdubied) [reply pkg/agents/team6/treaties.go]: good catch! we changed it

**Review by jzzheng22 — APPROVED** [review]
Thanks all :)

**Comment by MattScottEEE** [comment]
Will wait for Jaafar's review in the morning before merging, in case of any last min requests. Other than this, thanks everyone fro the great reviews <3

**Pull request merged**
